### PR TITLE
test: add Breadfund fuzz coverage (membership, withdraws, ratio, governance)

### LINKED ISSUE #16: Fuzzy Testing
Request to perform fuzzy testing on the Breadfunds project to identify unexpected behaviors or edge cases. Please outline the scope and methods for the fuzzy testing and document any issues uncovered during the process.

### PR DESCRIPTION
**Pull request by franrolotti**

### CODE REVIEW

File: test/invariants/fuzz/BreadfundFuzz_Bugs.t.sol
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Bugs.t.sol:1]: oh wow a whole set of tests on known bugs? what model created these so I can start using it

File: test/invariants/fuzz/README.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+* **`BreadfundFuzz_Soak_VariableMembers.t.sol`**
+  Large-scale stress tests:
+
+  * Random operations across many epochs.
+  * Variable member counts (up to \~25).
+  * Ensures no catastrophic state corruption under prolonged activity.
+
+---
+
+## Findings (Bug Spotlights)
```
Comment (bagelface) [inline test/invariants/fuzz/README.md]: We should take this section and make issues out of each one. That being said, I don't think this MD file needs to be included in this PR

File: test/invariants/fuzz/BreadfundFuzz_Bugs.t.sol
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Bugs.t.sol:1]: these tests don't appear to be fuzz tests. I think we can just copy these tests to the issues that are made for these bugs and get rid of this file

File: test/invariants/fuzz/BreadfundFuzzBase.t.sol [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  /**
+   * @dev Soft conservation check.
+   * Sums member withdrawables and asserts:
+   *   (contract token balance + amounts paid out via executed large requests) ≥ sum(withdrawables)
+   *
+   * Rationale: There is a known issue where executing large requests does not
+   * decrement withdrawables. To keep long soaks exploring state without failing
+   * on that bug, we add `executedOut` to the contract balance when checking.
+   *
+   * When the bug is fixed, prefer a strict variant in tests where no large
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzzBase.t.sol]: lets make note of this as well. we ideally aren't bringing in code that exists only to work around a known bug. need to make sure this doesn't get left in after the bug is fixed

File: test/invariants/fuzz/BreadfundFuzz_Create.t.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      if (cfg.consensusThreshold < 1) cfg.consensusThreshold = 1; // enforce lower bound in test input
+      cfg.autoThreshold = SAFE_AUTO_THRESHOLD + i * 1e15;         // drift to vary scenarios
+      cfg.members = _threeMembers();
+
+      // Every 5th: zero-address member should cause a revert on create.
+      // When spec defines the exact error, replace with the specific selector:
+      //   vm.expectRevert(IBreadfund.InvalidMemberAddress.selector);
+      if (i % 5 == 0) {
+        cfg.members[0] = address(0);
+        vm.expectRevert(); // IBreadfund.InvalidMemberAddress.selector (if used)
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Create.t.sol]: throw specific error

File: test/invariants/fuzz/BreadfundFuzz_Create.t.sol [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        continue;
+      }
+
+      // Every 9th: token not allowed → revert; then allow and retry (should succeed).
+      // When spec defines the exact error, replace with the specific selector:
+      //   vm.expectRevert(IBreadfund.TokenNotAllowed.selector);
+      if (i % 9 == 0) {
+        MockERC20 tmp = new MockERC20("Tmp", "TMP");
+        cfg.token = address(tmp);
+        vm.expectRevert(); // IBreadfund.TokenNotAllowed.selector (if used)
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Create.t.sol]: throw specific error

File: test/invariants/fuzz/BreadfundFuzz_Create.t.sol [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      uint256 id = breadfund.create(cfg);
+      assertEq(id, success, "id matches successes so far");
+      success++;
+    }
+
+    // Global property: nextId equals the count of successful creations.
+    assertEq(breadfund.nextId(), success, "nextId equals the number of successful creations");
+  }
+}
+// ────────────────────────────────────────────────────────────────────────────────
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Create.t.sol]: this comment not necessary

File: test/invariants/fuzz/BreadfundFuzz_Decommission.t.sol [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    assertEq(m1After - m1Before, w1 + split);
+    assertEq(m2After - m2Before, w2 + split);
+    assertEq(m3After - m3Before, w3 + split);
+
+    // Dust remainder (if any) stays in the contract
+    uint256 cBalAfter = token.balanceOf(address(breadfund));
+    assertEq(cBalAfter, remainder, "dust remainder retained");
+  }
+}
+/// ────────────────────────────────────────────────────────────────────────────────
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Decommission.t.sol]: this comment not necessary

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+      vm.expectRevert(IBreadfund.ExceedsSmallWithdrawalLimit.selector);
+      breadfund.withdraw(id, daysRequested);
+    }
+
+    // After advancing to new epoch, limit resets
+    vm.warp(block.timestamp + cfg.epochDuration + 1);
+    vm.prank(member1); breadfund.withdraw(id, daysRequested);
+  }
+}
+// ────────────────────────────────────────────────────────────────────────────────
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: this comment not necessary

File: test/invariants/fuzz/BreadfundFuzz_Ratio.t.sol [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+    _mintApproveLocal(localToken, member1, value + cfg.initialDeposit + cfg.fixedDeposit, address(localFund));
+    vm.prank(member1); localFund.deposit(idLocal, value);
+
+    uint256 contractHeld = localToken.balanceOf(address(localFund));
+    uint256 withdrawable = localFund.memberWithdrawableBalance(idLocal, member1);
+    assertGe(contractHeld, withdrawable, "conservation should hold for ratio <= 1");
+  }
+}
+/// ────────────────────────────────────────────────────────────────────────────────
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Ratio.t.sol]: this comment not necessary

File: test/invariants/fuzz/BreadfundFuzz_Create.t.sol [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      cfg.breadfundStart = block.timestamp + 1;
+      cfg.consensusThreshold = uint256((cBase + i) % 100);
+      if (cfg.consensusThreshold < 1) cfg.consensusThreshold = 1; // enforce lower bound in test input
+      cfg.autoThreshold = SAFE_AUTO_THRESHOLD + i * 1e15;         // drift to vary scenarios
+      cfg.members = _threeMembers();
+
+      // Every 5th: zero-address member should cause a revert on create.
+      // When spec defines the exact error, replace with the specific selector:
+      //   vm.expectRevert(IBreadfund.InvalidMemberAddress.selector);
+      if (i % 5 == 0) {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Create.t.sol]: weird approach. seems like it should just have its own test

File: test/invariants/fuzz/BreadfundFuzz_Create.t.sol [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        cfg.members[0] = address(0);
+        vm.expectRevert(); // IBreadfund.InvalidMemberAddress.selector (if used)
+        breadfund.create(cfg);
+        continue;
+      }
+
+      // Every 9th: token not allowed → revert; then allow and retry (should succeed).
+      // When spec defines the exact error, replace with the specific selector:
+      //   vm.expectRevert(IBreadfund.TokenNotAllowed.selector);
+      if (i % 9 == 0) {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Create.t.sol]: similarly should just have its own test

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+ *
+ * Properties checked:
+ *  - Deposits only once per epoch per member (`AlreadyDeposited`).
+ *  - Withdrawals respect daily contribution, thresholds, and per-epoch limits.
+ *  - Large withdrawals generate requests (and execute only after contest window).
+ *  - Soft conservation invariant (`contract balance + executed ≥ withdrawables`)
+ *    holds at each step.
+ *  - Epoch warp advances counters/reset correctly.
+ *
+ * Notes for reviewers:
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: this should not be included in comments

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+ *  - Some checks are unit-like but are *critical safety guards* (e.g.,
+ *    “already deposited this epoch”).
+ *  - After fixing BUG #2 (large exec not decrementing withdrawables),
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: need to make note of this somewhere besides comments here. include note in issue thats created for this bug

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+  /// -------------------------------------------------------------------------
+  /// Scenario: depositFor flow
+  /// Properties:
+  ///  - Non-member caller can deposit on behalf of a member.
+  ///  - Member flag is set for that epoch.
+  ///  - Second deposit by the member in the same epoch reverts.
+  ///  - depositFor with non-member target reverts.
+  /// -------------------------------------------------------------------------
+  function test_DepositFor_SetsMemberFlag_AndBlocksSecondDepositSameEpoch() public {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: not a fuzz test

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    vm.prank(caller);
+    vm.expectRevert(IBreadfund.NotMember.selector);
+    breadfund.depositFor(id, dep, notMember);
+  }
+
+  /// -------------------------------------------------------------------------
+  /// Scenario: guard deposit before fund start
+  /// Property: Deposits before `breadfundStart` must revert.
+  /// -------------------------------------------------------------------------
+  function test_Deposit_BeforeStart_Reverts() public {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: not a fuzz test

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+        seed = uint256(keccak256(abi.encodePacked(seed, e, k)));
+        address actor = _pick(defaultMembers, seed);
+        uint256 roll  = seed % 10;
+
+        uint256 epochIdx   = breadfund.getCurrentEpochIndex(id);
+        uint256 contrib    = breadfund.breadfundMemberContribute(id, actor);
+        uint256 withdrawableBefore = breadfund.memberWithdrawableBalance(id, actor);
+
+        // ── Case 1: Deposits ───────────────────────────────────────────────
+        if (roll <= 2) {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: personally don't like this "roll" approach. would rather just see the different cases have their own tests

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+
+  /// -------------------------------------------------------------------------
+  /// Scenario: small-withdraw limit fuzz
+  /// Properties:
+  ///  - Member can withdraw up to `smallWithdrawsLimit` times per epoch.
+  ///  - Extra attempts revert with `ExceedsSmallWithdrawalLimit`.
+  ///  - After advancing epoch, counter resets and withdrawals succeed again.
+  /// -------------------------------------------------------------------------
+  function testFuzz_SmallWithdrawsRespectLimit(
+    uint8 daysReqRaw, uint8 extraWithdrawsRaw, uint256 /*unused*/
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: why include this parameter if it's unused

File: test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+
+    for (uint256 j = 0; j < extraWithdraws; j++) {
+      vm.prank(member1);
+      vm.expectRevert(IBreadfund.ExceedsSmallWithdrawalLimit.selector);
+      breadfund.withdraw(id, daysRequested);
+    }
+
+    // After advancing to new epoch, limit resets
+    vm.warp(block.timestamp + cfg.epochDuration + 1);
+    vm.prank(member1); breadfund.withdraw(id, daysRequested);
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_DepositWithdraw.t.sol]: i didn't know you could have two statements on the same line like this. should def be a line break between them

File: test/invariants/fuzz/BreadfundFuzz_Ratio.t.sol [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+import {BreadfundFuzzBase} from "./BreadfundFuzzBase.t.sol";
+import {IBreadfund} from "src/interfaces/IBreadfund.sol";
+import {Breadfund} from "src/contracts/Breadfund.sol";
+import {MockERC20} from "test/mocks/MockERC20.sol";
+
+contract BreadfundFuzz_Ratio is BreadfundFuzzBase {
+  /// -------------------------------------------------------------------------
+  /// Scenario: Ratio > 1
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Ratio.t.sol]: this is also a bug test that should just be captured in an issue

File: test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+  ///  - After contest window elapses, an execution call succeeds.
+  ///  - Beneficiary's balance increases; request marked executed.
+  /// -------------------------------------------------------------------------
+  function testFuzz_LargeWithdrawal_RequestAndAutoExecute(
+    uint256 depositValueRaw, uint8 extraDaysRaw
+  ) public {
+    uint256 depositValue = bound(depositValueRaw, 5e18, 1e22);
+
+    IBreadfund.Breadfund memory cfg = safeCfg;
+    cfg.members = defaultMembers; cfg.ratio = 1; cfg.breadfundStart = block.timestamp;
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol]: line breaks between statements

File: test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+    uint256 id = breadfund.create(cfg);
+
+    uint256 totalNeeded = depositValue + cfg.initialDeposit + cfg.fixedDeposit;
+    _mintApprove(member1, totalNeeded, address(breadfund));
+    vm.prank(member1); breadfund.deposit(id, depositValue);
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol]: line breaks between statements

File: test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+  /// Scenario: Voting threshold semantics.
+  /// Properties:
+  ///  - With 4 members and threshold=50%, exactly two "yes" votes is NOT enough.
+  ///  - A third "yes" vote pushes above threshold and executes.
+  /// -------------------------------------------------------------------------
+  function test_Voting_ExactThreshold_NoExecute_ThenAboveExecutes() public {
+    address[] memory members = _makeMembers(4);
+
+    IBreadfund.Breadfund memory cfg = safeCfg;
+    cfg.members = members; cfg.minimumMembers = 2; cfg.maximumMembers = 4;
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol]: line breaks between statements

File: test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  /// -------------------------------------------------------------------------
+  /// Scenario: Voting windows and contest behavior.
+  /// Properties:
+  ///  - Only members may vote; double voting reverts.
+  ///  - Votes must be within `votingWindow`.
+  ///  - A contested request must not auto-execute on timeout.
+  /// -------------------------------------------------------------------------
+  function test_Voting_Windows_And_Contest_BlockAutoExec() public {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol]: line breaks between statements

Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol]: not a fuzz test

File: test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+    assertTrue(breadfund.isExecuted(reqId));
+  }
+
+  /// -------------------------------------------------------------------------
+  /// Scenario: Voting threshold semantics.
+  /// Properties:
+  ///  - With 4 members and threshold=50%, exactly two "yes" votes is NOT enough.
+  ///  - A third "yes" vote pushes above threshold and executes.
+  /// -------------------------------------------------------------------------
+  function test_Voting_ExactThreshold_NoExecute_ThenAboveExecutes() public {
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol]: not a fuzz test

File: test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  /// -------------------------------------------------------------------------
+  /// Scenario: Vote-heavy fuzz — random voting with bias until consensus or timeout.
+  /// Properties:
+  ///  - No catastrophic reverts while exploring varied member counts (3..20),
+  ///    consensus thresholds (1..99), and yes-vote biases (0..100%).
+  ///  - If consensus not reached by votes, allow post-window execution attempt.
+  /// -------------------------------------------------------------------------
+  function testFuzz_Voting_ConsensusOrTimeout(
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_RequestsVoting.t.sol:206]: line breaks between statements

File: test/invariants/fuzz/BreadfundFuzz_Soak_VariableMembers.t.sol [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+ */
+
+import {BreadfundFuzzBase} from "./BreadfundFuzzBase.t.sol";
+import {IBreadfund} from "src/interfaces/IBreadfund.sol";
+
+contract BreadfundFuzz_Soak_VariableMembers is BreadfundFuzzBase {
+  /// -------------------------------------------------------------------------
+  /// Scenario: Variable members, multiple epochs, randomized ops.
+  /// Ops mix:
+  ///  - roll ≤ 2: deposits (under r=1, withdrawable += v)
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Soak_VariableMembers.t.sol]: prefer separate tests to "rolling" for random scenarios

**Review by bagelface — CHANGES_REQUESTED** [review]
overall nice job, i think there are definitely some changes that should be made. seems like a whole set of bugs were discovered, so we definitely need to make issues for those.

**Comment by franrolotti** [comment]
Changes requested have been addressed. @bagelface

**Comment by RonTuretzky** [comment]
we need to rename `executeContestedWithdrawl` to `executeContestedWithdrawal`

**Comment by franrolotti** [comment]
@RonTuretzky executeContestedWithdrawal renamed here

File: test/invariants/fuzz/BreadfundFuzzBase.t.sol [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+    vm.stopPrank();
+  }
+
+  /// @dev Picks an address from `arr` using `seed` (modulo).
+  function _pick(address[] memory arr, uint256 seed) internal pure returns (address) {
+    return arr[seed % arr.length];
+  }
+
+
+
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzzBase.t.sol:184]: remove all these extra line breaks

File: test/invariants/fuzz/BreadfundFuzz_Decommission.t.sol [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+ * - Fund becomes decommissionable due to missed payments
+ *
+ * Properties checked:
+ *  - Contract is flagged decommissionable after a skipped epoch.
+ *  - Decommission distributes withdrawables + pro-rata share of remaining balance.
+ *  - Withdrawable balances are reset to 0 after decommission.
+ *  - The fund cannot be queried anymore (`NotCommissioned`).
+ *  - Dust remainder stays in the contract (explicit policy).
+ *
+ * Notes for reviewers:
```
Comment (bagelface) [inline test/invariants/fuzz/BreadfundFuzz_Decommission.t.sol]: this should not be included in comments

**Review by bagelface — APPROVED** [review]
I take issue with the way the comments are sort of overly descriptive and inconsistent throughout the tests. They will add detail that isn't needed. There's a "note to reviewers" which definitely needs to be removed.

I don't want to block merging based on comments, so I'm approving it, but I'd strongly encourage fixes the comments to keep only the necessary things (do we really need a header comment for each file that describes its "purpose"?) and make them consistent throughout the changes.

File: test/invariants/fuzz/BreadfundFuzz_Decommission.t.sol (hunk 27 above)
Comment (franrolotti) [reply test/invariants/fuzz/BreadfundFuzz_Decommission.t.sol]: Fixed in 85d1958

File: test/invariants/fuzz/BreadfundFuzzBase.t.sol (hunk 26 above)
Comment (franrolotti) [reply test/invariants/fuzz/BreadfundFuzzBase.t.sol:184]: Fixed in aaa5026
